refactor(compiler-cli): add the ability to remove imports from the import manager

crisbeto opened the pull request:
Extends the `ImportManager` to allow for imports to be removed. This will be useful in automated migrations.

devversion commented on packages/compiler-cli/src/ngtsc/translator/src/import_manager/import_manager.ts:
```diff
@@ -75,6 +75,7 @@ export class ImportManager
     }
   > = new Map();
 
+  private removedImports: Map<ts.SourceFile, Map<ModuleName, Set<string>>> = new Map();
```
Should this keep track of module names and their "export symbol + alias"? Maybe we can have a small JSDoc here.

(I assume not because we only allow deletion based on the export symbol)

crisbeto replied:
I intentionally wanted to deal only with the original symbol name, rather than the alias. Otherwise usage in migrations might get a bit annoying.

devversion replied:
:+1: should we add a small comment describing what the map maps to?

crisbeto replied:
Added a doc string.

devversion commented on packages/compiler-cli/src/ngtsc/translator/src/import_manager/import_manager.ts:
```diff
@@ -157,6 +166,33 @@ export class ImportManager
 
+  /**
+   * Marks all imported symbols with a specific name for removal.
+   * Call `addImport` to undo this operation.
```
Maybe add a note that all previous "expressions returned" via `addImport` can be invalidated that way? Seems like a notable caveat?

Also not sure if `addImport` can work to undo, because the returned`ts.Identifier`'s wouldn't be linked anymore, and a different alias may be used?

crisbeto replied:
The expressions added via `addImport` will still be valid, because the added imports are always being tracked. They get filtered out once at the end in `finalize`. I actually started off with untracking the imports when they're removed, but it ended up being kinda complicated and error-prone.

devversion replied:
I'm thinking of instances of e.g. `ts.Identifier`. Imagine you are doing the following:

```ts
myRef = addImport();

// use myRef.

// delete the import
// then re-add, but somewhere `myRef` from above is synthetically used!!
```

This can be problematic because TypeScript may use a "generated identifier" for `myRef` and rename it to something that doesn't cause conflicts. So the wrong "reference" is used somewhere.

crisbeto replied:
Maybe I'm misunderstanding, but I think that's what we have covered in https://github.com/angular/angular/pull/57110/files#diff-8ce9f28354ab2e84593261a78b2fe715c53ddcfb267a4ae9d8aae3cc281cd278R1035. When we generate the import and a unique name, it gets cached. `removeImport` never removes it from the cache so if it gets re-added, it'll preserve the same identifier.

devversion replied:
This is the part I'm referring to: https://github.com/angular/angular/blob/bd221c1d9134ada3dd5cf5d4a63ac1f4a4b3fb07/packages/compiler-cli/src/ngtsc/translator/src/import_manager/check_unique_identifier_name.ts#L51

TypeScript creates a `ts.Identifier`, but a special one. Whenever the AST nodes are printed to text form, TypeScript will do its own additional collision checking and when necessary, update the identifier (of both reference + specifier)

**but** I think I see whay you are saying? We won't create a new `ts.Identifier` but rather we just remove the data from `removedImports`?

crisbeto replied:
> but I think I see whay you are saying? We won't create a new ts.Identifier but rather we just remove the data from removedImports?

Yep exactly, the behavior is the same as before this change, except that it gets removed from the `removedImports`.

devversion commented on packages/compiler-cli/src/ngtsc/translator/src/import_manager/import_manager.ts:
```diff
@@ -255,10 +291,13 @@ export class ImportManager
     const updatedImportsResult = new Map<ts.NamedImports, ts.NamedImports>();
     const newImportsResult = new Map<string, ts.ImportDeclaration[]>();
+    const deletedImports = new Set<ts.ImportDeclaration>();
+    const importDeclarationsPerFile = new Map<ts.SourceFile, ts.ImportDeclaration[]>();
```
At this point I wish we could have this finalizing logic in a separate file.

crisbeto replied:
Hmm, I guess we could but it would require all the internal state of the manager to be passed around which might make it even messier.

devversion replied:
Fair enough. We can look into it as a follow-up, but not sure if you agree, but it seems quite large

crisbeto replied:
I think that it would be valuable to split up the method, I'm just not sure if it should be a separate function in a different file or just several private methods.

devversion replied:
Sounds good. We can look into it as a follow-up.

devversion reviewed (approved):
LGTM. Nice!

crisbeto committed:
refactor(compiler-cli): add the ability to remove imports from the import manager

Extends the `ImportManager` to allow for imports to be removed. This will be useful in automated migrations.

dylhunn commented:
This PR was merged into the repository by commit 2158b8dafb730cf43aaeb9b5ad05a67d7f706b6b.

The changes were merged into the following branches: main
